File: code/C/09.1.1_colorfulsoftlight/colorfulsoftlight.c
//
// Filename:    colorfulsoftlight.c
// Description: Use ADC module to read the voltage value of potentiometer.
// NOTE:        needs ADCDevice library from code/C/libraries/ADCDevice
//
////////////////////////////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////////////////////////////////
//
// Pre-processor directives
//
#include <wiringPi.h>
#include <softPwm.h>
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <ADCDevice.h>

#define  REDpin   3     // define 3 pins for RGBpin, set to wiringpi pin #'s
#define  GRNpin   2
#define  BLUpin   0

////////////////////////////////////////////////////////////////////////////////////////
//
// Declare Global ADCDevice variable
//
ADCDevice   *adc;

int32_t  main (void)
{
    ////////////////////////////////////////////////////////////////////////////////////
    //
    // Declare and initialize variables
    //
    int32_t  redval  = 0;
    int32_t  grnval  = 0;
    int32_t  bluval  = 0;

    ////////////////////////////////////////////////////////////////////////////////////
    //
    // Initialize the ADCDevice library
    //
    adc              = NULL;
    init_ADCDevice ();

    ////////////////////////////////////////////////////////////////////////////////////
    //
    // PWM: set the LED pins to PWM output, standard software range (0-100)
    //
    softPwmCreate (REDpin, 0, 100);
    softPwmCreate (GRNpin, 0, 100);
    softPwmCreate (BLUpin, 0, 100);

    fprintf (stdout, "Program is starting (CTRL-c to interrupt) ...\n");

    ////////////////////////////////////////////////////////////////////////////////////
    //
    // Main loop: proceed indefinitely until interruption
    //
    while (1)
    {
        ////////////////////////////////////////////////////////////////////////////////
        //
        // Read analog value of 3 potentiometers
        //
        redval       = adc -> analogRead (0);
        grnval       = adc -> analogRead (1);
        bluval       = adc -> analogRead (2);

        ////////////////////////////////////////////////////////////////////////////////
        //
        // Map the read value of the potentiometers into PWM value and output it
        //
        softPwmWrite (REDpin, redval * 100 / 255);
        softPwmWrite (GRNpin, grnval * 100 / 255);
        softPwmWrite (BLUpin, bluval * 100 / 255);

        ////////////////////////////////////////////////////////////////////////////////
        //
        // Display the current value for each color value
        //
        fprintf (stdout, "[ADC] red: %3d, green: %3d, blue: %3d \n", redval,
                                                                     grnval,
                                                                     bluval);
        delay (100);
    }

    return (0);
}

File: code/python/09.1.1_ColorfulSoftlight/ColorfulSoftlight.py
#!/usr/bin/env python3
#############################################################################
# Filename    : Softlight.py
# Description : Control RGBLED with Potentiometer 
# Author      : www.freenove.com
# modification: 2023/05/11
########################################################################

# Importing code libraries and code from ADCDevice file
from gpiozero import RGBLED
import time
from ADCDevice import *

#Setting the pins that will controll the amount of red green and blue as well as initializing the
#ADC variable
led = RGBLED(red=22, green=27, blue=17, active_high=False) # define the pins for R:GPIO22,G:GPIO27,B:GPIO17
#led = RGBLED(red="J8:15", green="J8:13", blue="J8:11") # according to BOARD Numbering define the pins for R:11,G:12,B:13 
adc = ADCDevice() # Define an ADCDevice class object

# Checks for if either the 0x48 or 0x4b code is detected and than will assign either the PCF or ADS
# Respectively to the adc variable. If not is detected it will print out an error and exit.
def setup():
    global adc
    if(adc.detectI2C(0x48)): # Detect the pcf8591.
        adc = PCF8591()
    elif(adc.detectI2C(0x4b)): # Detect the ads7830
        adc = ADS7830()
    else:
        print("No correct I2C address found, \n"
        "Please use command 'i2cdetect -y 1' to check the I2C address! \n"
        "Program Exit. \n");
        exit(-1)

# Creates a loop that while true will calculate the amount of red, green and blue should be
# Assigned to the led and then assigns it and prints out those values.    
def loop():
    while True:     
        value_Red = adc.analogRead(0)       # read ADC value of 3 potentiometers
        value_Green = adc.analogRead(1)
        value_Blue = adc.analogRead(2)
        led.red   =value_Red/255  # map the read value of potentiometers into PWM value and output it 
        led.green =value_Green/255
        led.blue  =value_Blue/255
        # print read ADC value
        print ('ADC Value value_Red: %d ,\tvlue_Green: %d ,\tvalue_Blue: %d'%(value_Red,value_Green,value_Blue))
        time.sleep(0.01)

# Garbage collection that will end the ads and led functions.
def destroy():
    adc.close()
    led.close()

# Checks if the current module is the main module and not imported. If so will print
# Program is starting do the setup above and try the loop above while checking for errors.
# If the keyboard interrupt is used it will cleanly stop the program and print program ending.    
if __name__ == '__main__': # Program entrance
    print ('Program is starting ... ')
    setup()
    try:
        loop()
    except KeyboardInterrupt: # Press ctrl-c to end the program.
        destroy()
        print("Ending program")
